### PR TITLE
JS: Compute global access paths and resolve JSDoc names

### PR DESCRIPTION
**Pull request by asger-semmle**
This computes the global access path referenced/aliased by a data flow node.

The main application is to augment the call graph and type tracking, but to avoid doing too many things at once, these are unchanged in this PR.

For now, the access paths are used to resolve JSDoc type names to their definitions, which is then used to implement jump-to-definition for JSDoc types. This forces evaluation of the new predicates, so performance can be measured before moving on to call graphs.

Evaluation on [big-apps](https://git.semmle.com/asger/dist-compare-reports/tree/canonical-name-defs_1563482687178) and [nightly](https://git.semmle.com/asger/dist-compare-reports/tree/canonical-name-defs_1563450663575) look reasonable

Commit-by-commit review is recommended.

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/SSA.qll [hunk 1]
```diff
@@ -521,6 +521,27 @@ class SsaExplicitDefinition extends SsaDefinition, TExplicitDef {
   ) {
     getDef().getLocation().hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn)
   }
+
+  /**
+   * Gets the data flow node representing the incoming value assigned at this definition,
+   * if any.
+   */
+  DataFlow::Node getRhsNode() {
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/SSA.qll]: This looks pretty similar to `defSourceNode` from `DataFlow.qll`. Could you comment on the differences?

File: javascript/ql/src/semmle/javascript/dataflow/DataFlow.qll [hunk 2]
```diff
@@ -161,6 +161,60 @@ module DataFlow {
...
+      )
+      or
+      // Definition of variable -> RHS of definition
+      exists(SsaExplicitDefinition def |
+        this = TSsaDefNode(def) and
+        result = def.getRhsNode()
+      )
+      or
+      // IIFE call -> return value of IIFE
+      // Note: not sound in case function falls over end and returns 'undefined'
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/dataflow/DataFlow.qll]: Should we do something about that? Perhaps it's time to introduce default-return nodes? (Doesn't have to be in this PR, of course.)

File: javascript/ql/src/semmle/javascript/dataflow/DataFlow.qll [hunk 3]
```diff
@@ -161,6 +161,60 @@ module DataFlow {
...
+      // Refinement of variable -> original definition of variable
+      exists(SsaRefinementNode refinement |
+        this.(DataFlow::SsaDefinitionNode).getSsaVariable() = refinement.getVariable() and
+        result.(DataFlow::SsaDefinitionNode).getSsaVariable() = refinement.getAnInput()
+      )
+      or
+      // Definition of variable -> RHS of definition
+      exists(SsaExplicitDefinition def |
+        this = TSsaDefNode(def) and
+        result = def.getRhsNode()
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/dataflow/DataFlow.qll]: In the presence of default values in destructuring assignments, a single definition can have multiple right-hand sides. We'd either have to `strictcount` that here, or introduce a new data-flow node that merges the right-hand sides.

File: javascript/ql/src/semmle/javascript/SSA.qll (hunk 1 above)
Comment (asger-semmle) [reply javascript/ql/src/semmle/javascript/SSA.qll]: As discussed, the only difference seems to be the inclusion of default values from destructuring reads. I've updated the PR to use this instead. I'd like to deal with the ambiguous RHS by changing the dataflow graph to ensure a unique RHS node actually exists.

File: javascript/ql/src/semmle/javascript/dataflow/DataFlow.qll (hunk 2 above)
Comment (asger-semmle) [reply javascript/ql/src/semmle/javascript/dataflow/DataFlow.qll]: I've loosened the contract for `getImmediatePredecessor` for now

File: javascript/ql/src/semmle/javascript/dataflow/DataFlow.qll [hunk 4]
```diff
@@ -161,6 +161,66 @@ module DataFlow {
...
+     * Gets the immediate predecessor of this node, if any.
+     *
+     * A node with an immediate predecessor can usually only have the value that flows
+     * into its from its immediate predecessor, currently with two exceptions:
+     *
+     * - An immediately-invoked function expression with a single return expression `e`
+     *   has `e` as its immediate predecessor, even if the function can fall over the
+     *   end and return `undefined`.
+     *
+     * - A destructuring property pattern with a default value has both the `PropRead`
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/dataflow/DataFlow.qll]:
```suggestion
     * - A destructuring property or element pattern with a default value has both the `PropRead`
```

File: javascript/ql/src/semmle/javascript/dataflow/DataFlow.qll [hunk 5]
```diff
@@ -1089,6 +1149,27 @@ module DataFlow {
...
+  /**
+   * INTERNAL. DO NOT USE.
+   *
+   * Gets the `PropRead` node corresponding to the value stored in the given
+   * binding pattern due to destructuring.
+   *
+   * For example, in `let { p: value } = f()`, the `value` pattern maps to a `PropRead`
+   * extracting the `p` property.
+   */
+  DataFlow::PropRead patternPropRead(BindingPattern value) {
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/dataflow/DataFlow.qll]: Can this be made `private`? It only seems to be used above in the same module.

Comment (asger-semmle) [reply javascript/ql/src/semmle/javascript/dataflow/DataFlow.qll]: Indeed, now that `getRhsNode` no longer uses it

**Pull request merged**